src/hotspot/share/opto/regmask.hpp

```diff
@@ -131,76 +401,85 @@ class RegMask {
     assert(valid_watermarks(), "post-condition");
   }
 
-  // Handy copying constructor
-  RegMask(RegMask *rm) {
-    _hwm = rm->_hwm;
-    _lwm = rm->_lwm;
-    for (unsigned i = 0; i < _RM_SIZE; i++) {
-      _RM_UP[i] = rm->_RM_UP[i];
-    }
+  // Construct an empty mask
+  RegMask(Arena* arena)
+      : _RM_UP(), _rm_size(_RM_SIZE), _offset(0), _all_stack(false),
+        _lwm(_RM_MAX), _hwm(0), _arena(arena) {
     assert(valid_watermarks(), "post-condition");
   }
-
-  // Construct an empty mask
-  RegMask() : _RM_UP(), _lwm(_RM_MAX), _hwm(0) {
+  RegMask() : RegMask(nullptr) {
     assert(valid_watermarks(), "post-condition");
   }
 
   // Construct a mask with a single bit
-  RegMask(OptoReg::Name reg) : RegMask() {
+  RegMask(OptoReg::Name reg,
+          Arena* arena
+          DEBUG_ONLY(COMMA bool orig_const = false))
+      : RegMask(arena) {
     Insert(reg);
+    DEBUG_ONLY(this->orig_const = orig_const;)
   }
+  RegMask(OptoReg::Name reg) : RegMask(reg, nullptr) {}
 
-  // Check for register being in mask
-  bool Member(OptoReg::Name reg) const {
-    assert(reg < CHUNK_SIZE, "");
-
-    unsigned r = (unsigned)reg;
-    return _RM_UP[r >> _LogWordBits] & (uintptr_t(1) << (r & _WordBitMask));
+  RegMask(const RegMask& rm, Arena* arena)
+      : _rm_size(_RM_SIZE), _offset(rm._offset), _arena(arena) {
+    _copy(rm);
   }
 
-  // The last bit in the register mask indicates that the mask should repeat
-  // indefinitely with ONE bits.  Returns TRUE if mask is infinite or
-  // unbounded in size.  Returns FALSE if mask is finite size.
-  bool is_AllStack() const {
-    return (_RM_UP[_RM_MAX] & (uintptr_t(1) << _WordBitMask)) != 0;
+  RegMask(const RegMask& rm) : RegMask(rm, nullptr) {}
+
+  RegMask& operator=(const RegMask& rm) {
+    _copy(rm);
+    return *this;
   }
 
-  void set_AllStack() {
-    _RM_UP[_RM_MAX] |= (uintptr_t(1) << _WordBitMask);
+  bool Member(OptoReg::Name reg) const {
+    reg = reg - offset_bits();
+    if (reg < 0) {
+      return false;
+    }
+    if (reg >= (int)rm_size_bits()) {
+      return is_AllStack();
+    }
+    unsigned int r = (unsigned int)reg;
+    return _rm_up(r >> _LogWordBits) & (uintptr_t(1) << (r & _WordBitMask));
   }
 
-  // Test for being a not-empty mask.
-  bool is_NotEmpty() const {
+  // Empty mask check. Ignores registers included through the all-stack flag.
+  bool is_Empty() const {
     assert(valid_watermarks(), "sanity");
-    uintptr_t tmp = 0;
     for (unsigned i = _lwm; i <= _hwm; i++) {
-      tmp |= _RM_UP[i];
+      if (_rm_up(i)) {
+        return false;
+      }
     }
-    return tmp;
+    return true;
   }
 
   // Find lowest-numbered register from mask, or BAD if mask is empty.
   OptoReg::Name find_first_elem() const {
     assert(valid_watermarks(), "sanity");
     for (unsigned i = _lwm; i <= _hwm; i++) {
-      uintptr_t bits = _RM_UP[i];
+      uintptr_t bits = _rm_up(i);
       if (bits) {
```

Comment:
This also looks like an implicit null check.

Reply:
Let's continue the discussion in the first comment (resolving this). Depending on what we decide, I'll fix all the (potential) implicit null checks.